### ISSUE TITLE
[BUG] KWArguments only passing one key:value pair

### ISSUE THREAD

**Issue by beardedeagle** [jenkinsci/saltstack-plugin#7]
When using the following settings (omitting user/pass type settings):

```
Client Interface: local
Target: tldweb*
Target Type: glob
Function: state.sls
Arguments: containers.update-rnvapi
KWArguments: 'pillar={image: "reg-cn-document-rest-api", tag: "2.0.6-SNAPSHOT.b177", env: "test", volume: "/opt/registrar/certs:/certs:ro", name: "rnv-api"}'
```

It is failing to pass the paramaters. Here is what I am seeing from a `salt-run state.event pretty=True`:

```
salt/job/20151124224350138890/ret/SERVERNAMEHERE	{
    "_stamp": "2015-11-25T05:43:50.384232",
    "cmd": "_return",
    "fun": "saltutil.find_job",
    "fun_args": [
        "20151124224310032167"
    ],
    "id": "SERVERNAMEHERE",
    "jid": "20151124224350138890",
    "retcode": 0,
    "return": {
        "arg": [
            "containers.update-rnvapi",
            {
                "'pillar": "{image: reg-cn-document-rest-api",
                "__kwarg__": true
            }
        ],
        "fun": "state.sls",
        "jid": "20151124224310032167",
        "pid": 12272,
        "ret": "",
        "tgt": "tldweb1*",
        "tgt_type": "glob",
        "user": "jenkins"
    },
    "success": true
}
```

Running this via the jenkins should be the equivilant of the following:

```
salt '*tldweb1*' state.sls containers.update-rnvapi 'pillar={image: "reg-cn-document-rest-api", tag: "2.0.6-SNAPSHOT.b177", env: "test", volume: "/opt/registrar/certs:/certs:ro, name: "rnv-api"}'
```

I am able to successfully run this locally on the salt-master via the salt command above, or from anywhere via a curl request so it isn't an issue with the state or the master. I was wondering if some light could be shed on this?

**Comment by mchugh19** [comment]
That's interesting. I would expect your syntax to function, so I'll have to look into it. 

I've got a working test case that does not quote the pillar values. Since your data does not include spaces, can you try removing the internal quotes to test?
Something like kwargs: "pillar={artifact: somevalue3, repo: othervalue4}"

When running that through test.arg, the pillar is not chopped at the quote boundary. 
...
"__pub_arg":     [
      "tests.pillar",
      {"pillar": "{artifact: somevalue3, repo: othervalue4}"}
    ],
    "pillar": "{artifact: somevalue3, repo: othervalue4}",
    "__pub_fun": "test.arg",
...

**Comment by beardedeagle** [comment]
That _does_ pass all kwargs, however that will not work for me:

```
Client Interface: local
Target: tldweb*
Target Type: glob
Function: state.sls
Arguments: containers.update-rnvapi
KWArguments: "pillar={image: reg-cn-document-rest-api, tag: 2.0.6-SNAPSHOT.b177, env: test, volume: /opt/registrar/certs:/certs:ro, name: rnv-api}"
```

Console output:

```
Response on state.sls containers.update-rnvapi for tldweb1*:
{"return": [{
  "SERVERNAMEHERE": "ERROR executing 'state.sls': Pillar data must be formatted as a dictionary",
  "SERVERNAMEHERE": "ERROR executing 'state.sls': Pillar data must be formatted as a dictionary"
}]}
Finished: SUCCESS
```

It is passing the KWArguments now though:

```
alt/job/20151125085458371964/new	{
    "_stamp": "2015-11-25T15:54:58.372527",
    "arg": [
        "containers.update-rnvapi",
        {
            "__kwarg__": true,
            "pillar": "{image: reg-cn-document-rest-api, tag: 2.0.6-SNAPSHOT.b177, env: test, volume: /opt/registrar/certs:/certs:ro, name: rnv-api}"
        }
    ],
    "fun": "state.sls",
    "jid": "20151125085458371964",
    "minions": [
        "SERVERNAMEHERE",
        "SERVERNAMEHERE"
    ],
    "tgt": "tldweb1*",
    "tgt_type": "glob",
    "user": "jenkins"
}
```

But it is failing because `volume: /opt/registrar/certs:/certs:ro` which needs to be `' '` in order to be interpreted as a string rather than a messed up key:value pair.

**Comment by mchugh19** [comment]
Alright, after playing around with this for a bit, it looks like there needs to be a cleanup in the quote handling. (also obvious from the existence of this conversation ;) )

As of now, I've successfully tested using multiple pillar values by adding them to the args line and wrapping the entire pillar object in double quotes. The values inside can optionally be wrapped in single quotes:
tests.pillar, "pillar={artifact: 'somevalue3', repo: 'othervalue:4'}"

So for a quick workaround, can you try your pillar line again, only add it to the args line and swap the quotes?
Arguments: "pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}"

**Comment by beardedeagle** [comment]
If I throw it all into Arguments:

```
Client Interface: local
Target: tldweb*
Target Type: glob
Function: state.sls
Arguments: containers.update-rnvapi "pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}"
````

Then I get the following console output:

```
Response on state.sls containers.update-rnvapi "pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}" for tldweb1*:
{"return": [{
  "SERVERNAMEHERE":   [
    "No matching sls found for \"containers.update-rnvapi pillar={image: 'reg-cn-document-rest-api'\" in env 'base'",
    "No matching sls found for \" tag: '2.0.6-SNAPSHOT.b177'\" in env 'base'",
    "No matching sls found for \" env: 'test'\" in env 'base'",
    "No matching sls found for \" volume: '/opt/registrar/certs:/certs:ro'\" in env 'base'",
    "No matching sls found for \" name: 'rnv-api'}\" in env 'base'"
  ],
  "SERVERNAMEHERE":   [
    "No matching sls found for \"containers.update-rnvapi pillar={image: 'reg-cn-document-rest-api'\" in env 'base'",
    "No matching sls found for \" tag: '2.0.6-SNAPSHOT.b177'\" in env 'base'",
    "No matching sls found for \" env: 'test'\" in env 'base'",
    "No matching sls found for \" volume: '/opt/registrar/certs:/certs:ro'\" in env 'base'",
    "No matching sls found for \" name: 'rnv-api'}\" in env 'base'"
  ]
}]}
Finished: SUCCESS
```

And the following from `salt-run state.event`:

```
salt/job/20151125095553233271/ret/SERVERNAMEHERE	{
    "_stamp": "2015-11-25T16:55:53.559959",
    "cmd": "_return",
    "fun": "state.sls",
    "fun_args": [
        "containers.update-rnvapi pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}"
    ],
    "id": "SERVERNAMEHERE",
    "jid": "20151125095553233271",
    "out": "highstate",
    "retcode": 1,
    "return": [
        "No matching sls found for \"containers.update-rnvapi pillar={image: 'reg-cn-document-rest-api'\" in env 'base'",
        "No matching sls found for \" tag: '2.0.6-SNAPSHOT.b177'\" in env 'base'",
        "No matching sls found for \" env: 'test'\" in env 'base'",
        "No matching sls found for \" volume: '/opt/registrar/certs:/certs:ro'\" in env 'base'",
        "No matching sls found for \" name: 'rnv-api'}\" in env 'base'"
    ],
    "success": true
}
salt/run/20151125095600603406/ret	{
    "_stamp": "2015-11-25T16:56:01.319110",
    "fun": "runner.jobs.list_job",
    "jid": "20151125095600603406",
    "return": {
        "Arguments": [
            "containers.update-rnvapi",
            {
                "'pillar": "{image: reg-cn-document-rest-api",
                "__kwarg__": true
            }
        ],
        "Function": " state.apply",
        "Minions": [
            "SERVERNAMEHERE",
            "SERVERNAMEHERE"
        ],
        "Result": {
            "SERVERNAMEHERE": {
                "out": "nested",
                "return": "' state.apply' is not available."
            },
            "SERVERNAMEHERE": {
                "out": "nested",
                "return": "' state.apply' is not available."
            }
        },
        "StartTime": "2015, Nov 25 09:40:27.797909",
        "Target": "*tldweb1*",
        "Target-type": "glob",
        "User": "jenkins",
        "jid": "20151125094027797909"
    },
    "success": true,
    "user": "jenkins"
}
```

Thought it is possible I am misinterpreting what you meant. I edited my previous comment to include what the quote handling should look like when ran successfully from command line as the jenkins user through pam auth.

**Comment by mchugh19** [comment]
In your arguments line, you'll need to separate with commas
Arguments: containers.update-rnvapi**,** "pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}"
(Bolding a single comma doesn't show up well. I just added it after your state: 
"containers.update-rnvapi,")

That should cause the salt plugin to generate something like the following json to be sent to the saltapi:
```
[{"client":"local","tgt":"salttesthost","expr_form":"list","fun":"state.sls","arg":["tests.pillar","pillar={artifact: 'somevalue3', repo: 'othervalue:4'}"]}]
```
If you'd like to validate the json, you can toggle its display with the debug checkbox in the saltapi section of the jenkins configuration page,

Salt should interpret that and parse the pillar:
```
Response on test.arg tests.pillar, "pillar={artifact: 'somevalue3', repo: 'othervalue:4'}" for salttesthost, stratos.kix:
[{"salttesthost": {
  "args": ["tests.pillar"],
  "kwargs":   {
    "__pub_user": "xian",
    "__pub_arg":     [
      "tests.pillar",
      "pillar={artifact: 'somevalue3', repo: 'othervalue:4'}"
    ],
    "pillar":     {
      "repo": "othervalue:4",
      "artifact": "somevalue3"
    },
    "__pub_fun": "test.arg",
    "__pub_jid": "20151125114007411052",
    "__pub_tgt": "salttesthost",
    "__pub_tgt_type": "list",
    "__pub_ret": ""
  }
}}]
```

**Comment by beardedeagle** [comment]
Ah, my bad. Totally blew past the comma in your response. That completely works and the state kicked off with all pillar data properly passed. This will do as a work around until the release for the reworked quote handling is released. Thank you sir, you are a gentleman and a scholar. I will leave this open since it is still technically an issue. Feel free to close it if you feel it should be.

**Comment by mchugh19** [comment]
Huzzah! When you get the time, can you try this with your original quoting? You'll still have to put the pillar info in the arguments field, but hopefully this resolves needing to wrap everything in double quotes instead of allowing single quotes. 
https://github.com/jenkinsci/saltstack-plugin/releases/download/v1.5.99-beta.2/saltstack.hpi

**Comment by mchugh19** [comment]
Scratch the previous version. I had to correct another part of the quote handling. This one should work as expected. (Still requires adding pillar data to arguments, but you can now use either single or double quotes as you wish)

https://github.com/jenkinsci/saltstack-plugin/releases/download/v1.5.99-beta.3/saltstack.hpi

**Comment by beardedeagle** [comment]
I'll get you results back as soon as I can.

**Comment by beardedeagle** [comment]
Apologies, life got in the way *AND* I had to set up a new jenkins master since the one we use at work is an older version and seems the beta3 release is not compatible with it. Running Jenkins 1.638 and saltstack-plugin v1.5.99-beta.3 I am able to successfully run the two different configs:
```
Client Interface: local
Target: tldweb*
Target Type: glob
Function: state.sls
Arguments: containers.update-rnvapi
KWArguments: "pillar={image: 'reg-cn-document-rest-api', tag: '2.0.6-SNAPSHOT.b177', env: 'test', volume: '/opt/registrar/certs:/certs:ro', name: 'rnv-api'}"
```

```
Client Interface: local
Target: tldweb*
Target Type: glob
Function: state.sls
Arguments: containers.update-rnvapi
KWArguments: 'pillar={image: "reg-cn-document-rest-api", tag: "2.0.6-SNAPSHOT.b177", env: "test", volume: "/opt/registrar/certs:/certs:ro", name: "rnv-api"}'
```

GREAT SUCCESS! Looks good to go. I will keep this open in case you would like to test anything else. Close it if you feel it should be.

**Comment by mchugh19** [comment]
Released with version 1.6
